src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -684,9 +690,13 @@ private function writeCollection($zval, $property, $collection, $addMethod, $rem
             $previousValue = false;
         }
 
-        foreach ($collection as $item) {
-            if (!$previousValue || !in_array($item, $previousValue, true)) {
-                $zval[self::VALUE]->{$addMethod}($item);
+        foreach ($collection as $key => $item) {
+            if ($secondKeyParameter) {
+                $zval[self::VALUE]->{$addMethod}($item, $key);
+            } else {
+                if (!$previousValue || !in_array($item, $previousValue, true)) {
```

Comment:
Could be simplified to `}  elseif (!$previousValue || !in_array($item, $previousValue, true)) {`